### ISSUE TITLE
handlers are not executed on all hosts when strategy free applied

### ISSUE THREAD

**Issue by tmekaumput** [ansible/ansible#31504]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

handlers

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = /home/jenkins/.ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Sep  5 2016, 02:30:38) [GCC 4.8.5 20150623 (Red Hat 4.8.5-9)]

```

##### OS / ENVIRONMENT

N/A

##### SUMMARY

Tasks in handlers block are not executed on all hosts when strategy **free** applied
The tasks in handlers block are executed OK on all hosts with linear strategy also including when serial batch process applied.

##### STEPS TO REPRODUCE

```yaml
- name: Test handler
  hosts: admin_host, app_hosts
  strategy: free
  tasks:
    - name: "Step #1"
      pause:
        seconds: "{{ 10 | random }}"

    - name: "Step #2"
      pause:
        seconds: "{{ 10 | random }}"
      notify: process_done

  handlers:
    - name: process_done
      debug:
        msg: "Hostname: {{ inventory_hostname }}"

```

##### EXPECTED RESULTS

```
PLAY [Test handlers - init] *************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************
ok: [localhost]

TASK [add_host] *************************************************************************************************************
changed: [localhost]

TASK [add_host] *************************************************************************************************************
changed: [localhost]

TASK [add_host] *************************************************************************************************************
changed: [localhost]

PLAY [Test handler] *********************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************
ok: [ofmw01]
ok: [ofmw03]
ok: [ofmw02]

TASK [Step #1] **************************************************************************************************************
changed: [ofmw03]
changed: [ofmw02]
changed: [ofmw01]

TASK [Step #2] **************************************************************************************************************
changed: [ofmw01]

RUNNING HANDLER [process_done] **********************************************************************************************

TASK [process_done] *********************************************************************************************************
ok: [ofmw01] => {
    "changed": false,
    "msg": "Hostname: ofmw01"
}

TASK [Step #2] **************************************************************************************************************
changed: [ofmw02]
changed: [ofmw03]

RUNNING HANDLER [process_done] **********************************************************************************************

TASK [process_done] *********************************************************************************************************
ok: [ofmw02] => {
    "changed": false,
    "msg": "Hostname: ofmw02"
}

RUNNING HANDLER [process_done] **********************************************************************************************

TASK [process_done] *********************************************************************************************************
ok: [ofmw03] => {
    "changed": false,
    "msg": "Hostname: ofmw03"
}


PLAY RECAP ******************************************************************************************************************
localhost                  : ok=4    changed=3    unreachable=0    failed=0
ofmw01                     : ok=4    changed=2    unreachable=0    failed=0
ofmw02                     : ok=4    changed=2    unreachable=0    failed=0
ofmw03                     : ok=4    changed=2    unreachable=0    failed=0
```

##### ACTUAL RESULTS

```
PLAY [Test handlers - init] *************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************
ok: [localhost]

TASK [add_host] *************************************************************************************************************
changed: [localhost]

TASK [add_host] *************************************************************************************************************
changed: [localhost]

TASK [add_host] *************************************************************************************************************
changed: [localhost]

PLAY [Test handler] *********************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************
ok: [ofmw01]
ok: [ofmw03]
ok: [ofmw02]

TASK [Step #1] **************************************************************************************************************
changed: [ofmw03]
changed: [ofmw02]
changed: [ofmw01]

TASK [Step #2] **************************************************************************************************************
changed: [ofmw01]

RUNNING HANDLER [process_done] **********************************************************************************************

TASK [process_done] *********************************************************************************************************
ok: [ofmw01] => {
    "changed": false,
    "msg": "Hostname: ofmw01"
}

TASK [Step #2] **************************************************************************************************************
changed: [ofmw02]
changed: [ofmw03]

PLAY RECAP ******************************************************************************************************************
localhost                  : ok=4    changed=3    unreachable=0    failed=0
ofmw01                     : ok=4    changed=2    unreachable=0    failed=0
ofmw02                     : ok=3    changed=2    unreachable=0    failed=0
ofmw03                     : ok=3    changed=2    unreachable=0    failed=0
```

```

```

**Comment by jborean93** [comment]
Are you able to verify your Ansible version again as I've tested this with both the latest devel and 2.3.0.0 and cannot replicate your issue. In fact, the `pause` module cannot be used with the free strategy and it will throw an error when you attempt to do so.

I worked around this by using the below, when running it, the handler is fired for both hosts as expected.

```
---
- hosts: localhost,Centos
  gather_facts: no
  strategy: free

  tasks:
  - name: "Step #1"
    command: hostname

  - name: "Step #2"
    command: hostname
    notify: process_done

  handlers:
    - name: process_done
      debug:
        msg: "Hostname: {{ inventory_hostname }}"
```

needs_info

**Comment by pablorf-dev** [comment]
This also affects me.

As @jborean93 said, pause doesn't work, but I rewrote the playbook to use shell and sleep as follows:

```
---
- name: Test handler
  hosts: all
  strategy: free
  tasks:
    - name: "Step #1"
      shell: "sleep {{ 10 | random }}"

    - name: "Step #2"
      shell: "sleep {{ 10 | random }}"
      notify: process_done

  handlers:
    - name: process_done
      debug:
        msg: "Hostname: {{ inventory_hostname }}"
```

And I can reproduce the bug:
```
$ ansible-playbook -c local -i 127.0.0.1,127.0.0.2 -Dvv /tmp/test.yml
ansible-playbook 2.4.1.0
  config file = None
  configured module search path = [u'/home/pablo/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible-playbook
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
No config file found; using defaults

PLAYBOOK: test.yml *************************************************************
1 plays in /tmp/test.yml

PLAY [Test handler] ************************************************************

TASK [Gathering Facts] *********************************************************
ok: [127.0.0.2]
META: ran handlers
ok: [127.0.0.1]
META: ran handlers

TASK [Step #1] *****************************************************************
task path: /tmp/test.yml:6
changed: [127.0.0.2] => {"changed": true, "cmd": "sleep 1", "delta": "0:00:01.002640", "end": "2017-11-10 09:40:24.897349", "failed": false, "rc": 0, "start": "2017-11-10 09:40:23.894709", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
changed: [127.0.0.1] => {"changed": true, "cmd": "sleep 3", "delta": "0:00:03.002202", "end": "2017-11-10 09:40:26.915542", "failed": false, "rc": 0, "start": "2017-11-10 09:40:23.913340", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
NOTIFIED HANDLER process_done

TASK [Step #2] *****************************************************************
task path: /tmp/test.yml:9
changed: [127.0.0.2] => {"changed": true, "cmd": "sleep 3", "delta": "0:00:03.002339", "end": "2017-11-10 09:40:28.042368", "failed": false, "rc": 0, "start": "2017-11-10 09:40:25.040029", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

RUNNING HANDLER [process_done] *************************************************

TASK [process_done] ************************************************************
task path: /tmp/test.yml:14
ok: [127.0.0.2] => {
    "msg": "Hostname: 127.0.0.2"
}
NOTIFIED HANDLER process_done

TASK [Step #2] *****************************************************************
task path: /tmp/test.yml:9
changed: [127.0.0.1] => {"changed": true, "cmd": "sleep 6", "delta": "0:00:06.002702", "end": "2017-11-10 09:40:33.090821", "failed": false, "rc": 0, "start": "2017-11-10 09:40:27.088119", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
META: ran handlers
META: ran handlers
META: ran handlers
META: ran handlers

PLAY RECAP *********************************************************************
127.0.0.1                  : ok=3    changed=2    unreachable=0    failed=0
127.0.0.2                  : ok=4    changed=2    unreachable=0    failed=0
```

**Comment by pablorf-dev** [comment]
After reviewing debug log, I believe it's a concurrency problem: notification of one host arrives just before the handler execution of the other just finish, wiping all notifications (check trace 'done running handlers').

The debug log is attached.

[handler.free.log.gz](https://github.com/ansible/ansible/files/1466161/handler.free.log.gz)

**Comment by pablorf-dev** [comment]
Check #32852 comments

**Comment by pablorf-dev** [comment]
I provide another example that is closer to my scenario (it involves multiple wait for host).

##### STEPS TO REPRODUCE
Use the following playbook:
```
---
- name: Test handler
  hosts: 127.0.0.*
  connection: local
  strategy: free
  vars:
    timeout:
      '127.0.0.1': 5
      '127.0.0.2': 70
  tasks:
    - name: "Step #1"
      shell: "sleep {{ timeout[inventory_hostname] }}"
      notify: stage1

    - name: "Step #2"
      shell: "sleep {{ 5 | random }}"

  handlers:
    - name: "Stage 1 - Step #1"
      shell: "sleep {{ item }}"
      listen: stage1
      with_items:
        - "{{ 5 | random }}"
        - "{{ 5 | random }}"

    - name: "Stage 1 - Step #2"
      shell: "sleep {{ 5 | random }}"
      listen: stage1
      notify: stage2

    - name: "Stage 2"
      debug:
        msg: "{{ inventory_hostname }} second stage"
      listen: stage2
```

Command line:
```
ansible-playbook -i 127.0.0.1,127.0.0.2 -Dvv -c local /tmp/free.test.yml
```
##### EXPECTED RESULT
```
ansible-playbook 2.4.1.0 (detached HEAD e946c4b3b7) last updated 2017/11/14 08:53:50 (GMT +200)
  config file = None
  configured module search path = [u'/home/pablo/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/pablo/ansible/lib/ansible
  executable location = /home/pablo/ansible/bin/ansible-playbook
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
No config file found; using defaults

PLAYBOOK: free.test.yml *********************************************************************************
1 plays in /tmp/free.test.yml

PLAY [Test handler] *************************************************************************************

TASK [Gathering Facts] **********************************************************************************
ok: [127.0.0.1]
META: ran handlers
ok: [127.0.0.2]
META: ran handlers
NOTIFIED HANDLER Stage 1 - Step #1
NOTIFIED HANDLER Stage 1 - Step #2

TASK [Step #1] ******************************************************************************************
task path: /tmp/free.test.yml:6
changed: [127.0.0.2] => {"changed": true, "cmd": "sleep 3", "delta": "0:00:03.002439", "end": "2017-11-14 10:56:11.769406", "failed": false, "rc": 0, "start": "2017-11-14 10:56:08.766967", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

TASK [Step #2] ******************************************************************************************
task path: /tmp/free.test.yml:10
changed: [127.0.0.2] => {"changed": true, "cmd": "sleep 7", "delta": "0:00:07.002994", "end": "2017-11-14 10:56:18.910867", "failed": false, "rc": 0, "start": "2017-11-14 10:56:11.907873", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

RUNNING HANDLER [Stage 1 - Step #1] *********************************************************************
changed: [127.0.0.2] => (item=1) => {"changed": true, "cmd": "sleep 1", "delta": "0:00:01.003190", "end": "2017-11-14 10:56:20.079050", "failed": false, "item": "1", "rc": 0, "start": "2017-11-14 10:56:19.075860", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
changed: [127.0.0.2] => (item=1) => {"changed": true, "cmd": "sleep 1", "delta": "0:00:01.002707", "end": "2017-11-14 10:56:21.232590", "failed": false, "item": "1", "rc": 0, "start": "2017-11-14 10:56:20.229883", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

TASK [Stage 1 - Step #1] ********************************************************************************
task path: /tmp/free.test.yml:14

RUNNING HANDLER [Stage 1 - Step #2] *********************************************************************
NOTIFIED HANDLER Stage 2

TASK [Stage 1 - Step #2] ********************************************************************************
task path: /tmp/free.test.yml:21
changed: [127.0.0.2] => {"changed": true, "cmd": "sleep 9", "delta": "0:00:09.002316", "end": "2017-11-14 10:56:30.937223", "failed": false, "rc": 0, "start": "2017-11-14 10:56:21.934907", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

RUNNING HANDLER [Stage 2] *******************************************************************************

TASK [Stage 2] ******************************************************************************************
task path: /tmp/free.test.yml:26
ok: [127.0.0.2] => {
    "msg": "127.0.0.2 second stage"
}

NOTIFIED HANDLER Stage 1 - Step #1
NOTIFIED HANDLER Stage 1 - Step #2

TASK [Step #1] ******************************************************************************************
task path: /tmp/free.test.yml:6
changed: [127.0.0.1] => {"changed": true, "cmd": "sleep 76", "delta": "0:01:16.003352", "end": "2017-11-14 10:57:24.769803", "failed": false, "rc": 0, "start": "2017-11-14 10:56:08.766451", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

TASK [Step #2] ******************************************************************************************
task path: /tmp/free.test.yml:10
changed: [127.0.0.1] => {"changed": true, "cmd": "sleep 7", "delta": "0:00:07.003360", "end": "2017-11-14 10:57:32.132504", "failed": false, "rc": 0, "start": "2017-11-14 10:57:25.129144", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
META: ran handlers
META: ran handlers

RUNNING HANDLER [Stage 1 - Step #1] *********************************************************************
changed: [127.0.0.1] => (item=1) => {"changed": true, "cmd": "sleep 1", "delta": "0:00:01.003190", "end": "2017-11-14 10:57:34.079050", "failed": false, "item": "1", "rc": 0, "start": "2017-11-14 10:57:33.075860", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
changed: [127.0.0.1] => (item=1) => {"changed": true, "cmd": "sleep 1", "delta": "0:00:01.002707", "end": "2017-11-14 10:57:35.232590", "failed": false, "item": "1", "rc": 0, "start": "2017-11-14 10:57:34.229883", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

RUNNING HANDLER [Stage 1 - Step #2] *********************************************************************
NOTIFIED HANDLER Stage 2

TASK [Stage 1 - Step #2] ********************************************************************************
task path: /tmp/free.test.yml:21
changed: [127.0.0.1] => {"changed": true, "cmd": "sleep 5", "delta": "0:00:05.002650", "end": "2017-11-14 10:57:40.945462", "failed": false, "rc": 0, "start": "2017-11-14 10:57:35.942812", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

RUNNING HANDLER [Stage 2] *******************************************************************************

TASK [Stage 2] ******************************************************************************************
task path: /tmp/free.test.yml:26
ok: [127.0.0.1] => {
    "msg": "127.0.0.1 second stage"
}
META: ran handlers
META: ran handlers

PLAY RECAP **********************************************************************************************
127.0.0.1                  : ok=6    changed=4    unreachable=0    failed=0
127.0.0.2                  : ok=6    changed=4    unreachable=0    failed=0
```
##### ACTUAL RESULT
```
ansible-playbook 2.4.1.0 (detached HEAD e946c4b3b7) last updated 2017/11/14 08:53:50 (GMT +200)
  config file = None
  configured module search path = [u'/home/pablo/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/pablo/ansible/lib/ansible
  executable location = /home/pablo/ansible/bin/ansible-playbook
  python version = 2.7.13 (default, Jan 19 2017, 14:48:08) [GCC 6.3.0 20170118]
No config file found; using defaults

PLAYBOOK: free.test.yml *********************************************************************************
1 plays in /tmp/free.test.yml

PLAY [Test handler] *************************************************************************************

TASK [Gathering Facts] **********************************************************************************
ok: [127.0.0.1]
META: ran handlers
ok: [127.0.0.2]
META: ran handlers
NOTIFIED HANDLER Stage 1 - Step #1
NOTIFIED HANDLER Stage 1 - Step #2

TASK [Step #1] ******************************************************************************************
task path: /tmp/free.test.yml:6
changed: [127.0.0.2] => {"changed": true, "cmd": "sleep 3", "delta": "0:00:03.002439", "end": "2017-11-14 10:56:11.769406", "failed": false, "rc": 0, "start": "2017-11-14 10:56:08.766967", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

TASK [Step #2] ******************************************************************************************
task path: /tmp/free.test.yml:10
changed: [127.0.0.2] => {"changed": true, "cmd": "sleep 7", "delta": "0:00:07.002994", "end": "2017-11-14 10:56:18.910867", "failed": false, "rc": 0, "start": "2017-11-14 10:56:11.907873", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

RUNNING HANDLER [Stage 1 - Step #1] *********************************************************************
changed: [127.0.0.2] => (item=1) => {"changed": true, "cmd": "sleep 1", "delta": "0:00:01.003190", "end": "2017-11-14 10:56:20.079050", "failed": false, "item": "1", "rc": 0, "start": "2017-11-14 10:56:19.075860", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
changed: [127.0.0.2] => (item=1) => {"changed": true, "cmd": "sleep 1", "delta": "0:00:01.002707", "end": "2017-11-14 10:56:21.232590", "failed": false, "item": "1", "rc": 0, "start": "2017-11-14 10:56:20.229883", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

TASK [Stage 1 - Step #1] ********************************************************************************
task path: /tmp/free.test.yml:14


(Here gets frozen)


NOTIFIED HANDLER Stage 1 - Step #1
NOTIFIED HANDLER Stage 1 - Step #2

TASK [Step #1] ******************************************************************************************
task path: /tmp/free.test.yml:6
changed: [127.0.0.1] => {"changed": true, "cmd": "sleep 76", "delta": "0:01:16.003352", "end": "2017-11-14 10:57:24.769803", "failed": false, "rc": 0, "start": "2017-11-14 10:56:08.766451", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

RUNNING HANDLER [Stage 1 - Step #2] *********************************************************************
NOTIFIED HANDLER Stage 2

TASK [Stage 1 - Step #2] ********************************************************************************
task path: /tmp/free.test.yml:21
changed: [127.0.0.1] => {"changed": true, "cmd": "sleep 5", "delta": "0:00:05.002650", "end": "2017-11-14 10:57:29.945462", "failed": false, "rc": 0, "start": "2017-11-14 10:57:24.942812", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
NOTIFIED HANDLER Stage 2
changed: [127.0.0.2] => {"changed": true, "cmd": "sleep 9", "delta": "0:00:09.002316", "end": "2017-11-14 10:57:33.937223", "failed": false, "rc": 0, "start": "2017-11-14 10:57:24.934907", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

RUNNING HANDLER [Stage 2] *******************************************************************************

TASK [Stage 2] ******************************************************************************************
task path: /tmp/free.test.yml:26
ok: [127.0.0.1] => {
    "msg": "127.0.0.1 second stage"
}
ok: [127.0.0.2] => {
    "msg": "127.0.0.2 second stage"
}
META: ran handlers
META: ran handlers

TASK [Step #2] ******************************************************************************************
task path: /tmp/free.test.yml:10
changed: [127.0.0.1] => {"changed": true, "cmd": "sleep 7", "delta": "0:00:07.003360", "end": "2017-11-14 10:57:41.132504", "failed": false, "rc": 0, "start": "2017-11-14 10:57:34.129144", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
META: ran handlers
META: ran handlers

PLAY RECAP **********************************************************************************************
127.0.0.1                  : ok=5    changed=3    unreachable=0    failed=0
127.0.0.2                  : ok=6    changed=4    unreachable=0    failed=0
```

**Comment by pablorf-dev** [comment]
After trying to understand the free strategy, I identified several points contributing to this issue:

1. The [following line](https://github.com/ansible/ansible/blob/v2.4.1.0-1/lib/ansible/plugins/strategy/__init__.py#L778) cause the freeze on free strategy with a handler on a live host and with other ones blocked: 
```
host_results = self._wait_on_pending_results(iterator)
```

2. [This one](https://github.com/ansible/ansible/blob/v2.4.1.0-1/lib/ansible/plugins/strategy/__init__.py#L820) causes a race condition when a notification of another host arrives while it is being executed:
```
self._notified_handlers[handler._uuid] = []
```

3. [Run handler](https://github.com/ansible/ansible/blob/v2.4.1.0-1/lib/ansible/plugins/strategy/__init__.py#L868) without using `target_host` provided causes the execution of another hosts' handlers when they still haven't finished with previous tasks executed after notification:
```
elif meta_action == 'flush_handlers':
    self.run_handlers(iterator, play_context)
```

I've been able to hotfix 2 and 3, but the 1 require to wait only for hosts that are executing the present handler, instead of all of them, and may require a whole rewrite of the functions `_process_pending_results` or `_wait_on_pending_results`, or even rewrite the structure of the metadata stored while queueing tasks in order to easily identify which tasks are handlers and if they are reached the point after they can be executed.

I can provide my modifications, but I think that fix this require a deeper knowledge of the code.

Any suggestions?

**Comment by bcoca** [comment]
so i couldn't reproduce this initially, but it was mostly due to lack of 'differences' in my tests hosts, the following generally reproduces the problem on almost every run.

```yaml
- hosts: all
  gather_facts: False
  strategy: free
  tasks:
    - wait_for: timeout={{5|random}}
      delegate_to: localhost
      changed_when: True
      notify: trip

  handlers:
    - debug: var=inventory_hostname
      name: trip
```

**Comment by pablorf-dev** [comment]
I've provided another pull request that seems to fix all the points I commented previously.

@bcoca The problem of your PR is that only fixes the wiping of all hosts, that actually fix this issue, but you may execute handlers unordered and get blocked by other tasks.

**Comment by bcoca** [comment]
@pablorf-dev agreed i had not accounted for 'premature handler running' which the other Issue does point to.

**Comment by AidasK** [comment]
affects ansible 2.4.2.0 too

**Comment by Nothing4You** [comment]
ansible 2.5.0 still affected, @bcoca's example consistenly only runs the handler on a single seemingly random host.

applying the patch from #33011 consistently runs the handler on every host.

**Comment by timesking** [comment]
ansible 2.5.0 still affected

**Comment by BradFelmey** [comment]
Behavior observable in Ansible 2.5.3.

**Comment by insideClaw** [comment]
Problem confirmed in 2.5.2. A task with a restart service handler fails when the following task (that requires the service to have finished restarting) fails. In other words, task # 2 fails because task # 1 completes even though the handler it invokes hasn't finished.
